sdk/src/wallet/migration/mod.rs

```diff
@@ -45,97 +75,71 @@ pub(crate) trait Migration {
 
-    #[cfg(feature = "storage")]
-    async fn migrate_storage(&self, storage: &super::storage::Storage) -> Result<()>;
-
-    #[cfg(feature = "stronghold")]
-    async fn migrate_backup(&self, storage: &crate::client::stronghold::StrongholdAdapter) -> Result<()>;
+    async fn migrate(&self, storage: &S) -> Result<()>;
 }
 
 #[async_trait]
-impl<T: Migration + Send + Sync> DynMigration for T {
+impl<S: StorageAdapter, T: Migration<S> + Send + Sync> DynMigration<S> for T
+where
+    crate::wallet::Error: From<S::Error>,
+    S::Error: From<serde_json::Error>,
+{
     fn version(&self) -> MigrationVersion {
         T::version()
     }
 
-    #[cfg(feature = "storage")]
-    async fn migrate_storage(&self, storage: &super::storage::Storage) -> Result<()> {
+    async fn migrate(&self, storage: &S) -> Result<()> {
         let version = self.version();
         log::info!("Migrating to version {}", version);
-        T::migrate_storage(storage).await?;
-        storage.set(MIGRATION_VERSION_KEY, version).await?;
-        Ok(())
-    }
-
-    #[cfg(feature = "stronghold")]
-    async fn migrate_backup(&self, storage: &crate::client::stronghold::StrongholdAdapter) -> Result<()> {
-        use crate::client::storage::StorageProvider;
-
-        let version = self.version();
-        log::info!("Migrating backup to version {}", version);
-        T::migrate_backup(storage).await?;
-        storage
-            .insert(
-                MIGRATION_VERSION_KEY.as_bytes(),
-                serde_json::to_string(&version)?.as_bytes(),
-            )
-            .await?;
+        T::migrate(storage).await?;
+        storage.set(MIGRATION_VERSION_KEY, &version).await?;
         Ok(())
     }
 }
 
-#[cfg(feature = "storage")]
-pub async fn migrate_storage(storage: &super::storage::Storage) -> Result<()> {
+pub async fn migrate<S: 'static + StorageAdapter>(storage: &S) -> Result<()>
+where
+    crate::wallet::Error: From<S::Error>,
+    S::Error: From<serde_json::Error>,
+{
     let last_migration = storage.get::<MigrationVersion>(MIGRATION_VERSION_KEY).await?;
     for migration in migrations(last_migration)? {
-        migration.migrate_storage(storage).await?;
+        migration.migrate(storage).await?;
     }
     Ok(())
 }
 
-#[cfg(feature = "stronghold")]
-pub async fn migrate_backup(storage: &crate::client::stronghold::StrongholdAdapter) -> Result<()> {
-    use crate::client::storage::StorageProvider;
-
-    let last_migration = storage
-        .get(MIGRATION_VERSION_KEY.as_bytes())
-        .await?
-        .map(|bytes| serde_json::from_slice::<MigrationVersion>(&bytes))
-        .transpose()?;
-    for migration in migrations(last_migration)? {
-        migration.migrate_backup(storage).await?;
+fn migrations<S: 'static + StorageAdapter>(
+    mut last_migration: Option<MigrationVersion>,
+) -> Result<Vec<&'static dyn DynMigration<S>>> {
+    let migrations = MIGRATIONS
+        .get::<HashMap<Option<usize>, &'static dyn DynMigration<S>>>()
```

Comment:
How does this differentiate between storage and stronghold migrations?
Also can you please explain why we need own migration versions for the storage and stronghold and can't just have them combined as before?

Reply:
The generic parameter differentiates them by type.

Splitting them is just cleaner really. Better usage of feature gates and extensibility if we need to change it in the future. Plus, we can migrate one and not the other, if the situation calls for it.

Reply:
> The generic parameter differentiates them by type.

Will this really not just get the first entry that matches `'static + StorageAdapter`, but the exact type that was provided there?

Reply:
This generic is the important bit: ```&'static dyn DynMigration<S>>```